### PR TITLE
forgot password functionality

### PR DESCRIPTION
**Pull request by theandrewmo**
### Description
This pull request introduces a new feature that allows users to reset their passwords in case they forget them or need to change them for security reasons. The implementation includes the necessary backend and frontend changes to provide a seamless password reset experience for our users.

### Changes Made
- Added new templates and routes to handle the password reset flow.
- Created a reset route to handle the password reset requests.
- Updated user model to store the password reset tokens.
- Added a new view to collect user email for initiating the password reset process.
- Developed the backend logic to generate and validate password reset tokens.
- Integrated email notifications to send password reset instructions to users.
- Implemented frontend validation and error handling for the password reset forms.

### Benefits
- Improved user experience: Users can now easily reset their passwords
- Enhanced security: The password reset tokens are securely generated and validated, ensuring that only the intended user can reset their password.

### Testing Done
Created unit tests for the new password reset functionality.
Performed manual testing of the password reset flow to verify its functionality and responsiveness.

### Screenshots / Demo

https://github.com/theandrewmo/hoppyhour/assets/106348010/edf67dda-a816-4db4-b450-3a9bde1ee6c9

### Additional Notes
Thank you for reviewing this pull request. Any feedback and suggestions are greatly appreciated.

### CODE REVIEW

File: app.py [hunk 1]
```diff
@@ -1,9 +1,10 @@
 import os, requests, json
 
 from flask import Flask, render_template, request, flash, redirect, session, g 
+from flask_mail import Mail, Message
```
Comment (akhayoon) [inline app.py:4]: let's make sure this works on Render. You might have to use an email service of some kind.

File: app.py [hunk 2]
```diff
@@ -141,6 +146,54 @@ def get_user():
...
+    form = ForgotPasswordForm()
+    if request.method == 'POST':
+        email = request.form['email']
+
+        user = User.query.filter_by(email=email).first()
+        if user:
+            token = generate_reset_token(user)
+
+            msg = Message('Hello, Here is your reset token', sender=mail_username, recipients=[user.email])
+            msg.body = f"http://127.0.0.1:5000/reset_password/{token}"
```
Comment (akhayoon) [inline app.py:165]: This URL needs to be based on the environment

File: templates/reset_password.html [hunk 3]
```diff
@@ -0,0 +1,22 @@
+{% extends 'base.html' %}
+
+{% block content %}
+
+<div class="row justify-content-center"> 
+    <div class="col-md-7 text-center">
+        <h2 class="mt-4">Enter new password</h2>
+        <form method="POST">
```
Comment (akhayoon) [inline templates/reset_password.html:8]: An alternative that has to be mentioned is to do the call async. Avoiding any page refreshes and you can confirm password matching before going to the server.

File: test_routes.py [hunk 4]
```diff
@@ -165,6 +165,69 @@ def test_edit(self):
             self.assertIsNotNone(updated_user)
             self.assertEqual(updated_user.fav_type, 'nano')
 
+
+    def test_forgot_password(self):
```
Comment (akhayoon) [inline test_routes.py:167]: It's more verbose but try to make the test cases more broken down.

**Review by akhayoon — APPROVED** [review]
LGTM

File: app.py (hunk 1 above)
Comment (theandrewmo) [reply app.py:4]: all set

File: app.py (hunk 2 above)
Comment (theandrewmo) [reply app.py:165]: done

File: templates/reset_password.html (hunk 3 above)
Comment (theandrewmo) [reply templates/reset_password.html:8]: done

File: test_routes.py (hunk 4 above)
Comment (theandrewmo) [reply test_routes.py:167]: done

**Pull request merged**